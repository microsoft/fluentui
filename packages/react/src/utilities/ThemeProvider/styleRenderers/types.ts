import { IFontFace, IKeyframes } from '@fluentui/merge-styles';

/* eslint-disable @typescript-eslint/naming-convention */

type StyleRendererOptions = {
  rtl?: boolean;
  targetWindow: Window | undefined;
};

export interface StyleRenderer {
  /**
   * Expected to initialize or re-initialize the renderer. Primarily for testing purposes.
   */
  reset: () => void;

  /**
   * Returns a unique id for the renderer; used as part of the cache key when determining if new
   * styles need to be rendered.
   */
  getId: () => number;

  /**
   * Renders a stylesheet and returns the map of key to class name.
   */
  renderStyles: <TRuleSet>(ruleSet: TRuleSet, options: StyleRendererOptions) => { [key in keyof TRuleSet]: string };

  /**
   * Renders keyframes and returns the keyframe name.
   */
  renderKeyframes: (keyframes: IKeyframes, options: StyleRendererOptions) => string;

  /**
   * Renders fontfaces.
   */
  renderFontFace: (fontFace: IFontFace, options: StyleRendererOptions) => void;
}
