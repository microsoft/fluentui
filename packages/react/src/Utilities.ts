import './version';
export * from '@fluentui/utilities';
