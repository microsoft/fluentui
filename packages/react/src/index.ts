export * from './ActivityItem';
export * from './Autofill';
export * from './Announced';
export * from './Breadcrumb';
export * from './Button';
export * from './ButtonGrid';
export * from './Calendar';
export * from './Callout';
export * from './Check';
export * from './Checkbox';
export * from './ChoiceGroup';
// export * from './ChoiceGroupOption'; // exported by ChoiceGroup
export * from './Coachmark';
export * from './Color';
export * from './ColorPicker';
export * from './ComboBox';
export * from './CommandBar';
export * from './ContextualMenu';
export * from './DatePicker';
export * from './DateTimeUtilities';
export * from './DetailsList';
export * from './Dialog';
export * from './Divider';
export * from './DocumentCard';
export * from './DragDrop';
export * from './Dropdown';
export * from './ExtendedPicker';
export * from './Fabric';
export * from './Facepile';
export * from './FloatingPicker';
export * from './FocusTrapZone';
export * from './FocusZone';
export * from './Grid';
export * from './GroupedList';
export * from './HoverCard';
export * from './Icon';
export * from './Icons';
export * from './Image';
export * from './Keytips';
export * from './Keytip';
export * from './KeytipData';
export * from './KeytipLayer';
export * from './Label';
export * from './Layer';
export * from './Link';
export * from './List';
export * from './MarqueeSelection';
export * from './MessageBar';
export * from './Modal';
export * from './Nav';
export * from './OverflowSet';
export * from './Overlay';
export * from './Panel';
export * from './Persona';
export * from './PersonaCoin';
// export * from './PersonaPresence'; (Exported as part of Persona)
export * from './Pickers';
export * from './Pivot';
export * from './Popup';
export * from './Positioning';
export * from './PositioningContainer';
export * from './ProgressIndicator';
export * from './Rating';
export * from './ResizeGroup';
export * from './ResponsiveMode';
export * from './ScrollablePane';
export * from './SearchBox';
export * from './SelectableOption';
export * from './SelectedItemsList';
export * from './Selection';
export * from './Separator';
export * from './Shimmer';
export * from './ShimmeredDetailsList';
export * from './Slider';
export * from './SpinButton';
export * from './Spinner';
export * from './Stack';
export * from './Sticky';
export * from './Styling';
export * from './SwatchColorPicker';
export * from './TeachingBubble';
export * from './Text';
export * from './TextField';
export * from './Theme';
export * from './ThemeGenerator';
export * from './TimePicker';
export * from './Toggle';
export * from './Tooltip';
export * from './Utilities';
export * from './WeeklyDayPicker';
export * from './WindowProvider';

import './version';
