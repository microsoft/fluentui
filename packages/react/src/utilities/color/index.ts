export * from './colors';
export * from './shades';
export * from './consts';
export * from './interfaces';
export * from './cssColor';
export * from './rgb2hex';
export * from './clamp';
export * from './hsl2rgb';
export * from './hsl2hsv';
export * from './hsv2rgb';
export * from './hsv2hex';
export * from './rgb2hsv';
export * from './hsv2hsl';
export * from './getColorFromString';
export * from './getColorFromRGBA';
export * from './getColorFromHSV';
export * from './getFullColorString';
export * from './updateSV';
export * from './updateH';
export * from './updateRGB';
export * from './getColorFromString';
export * from './updateA';
export * from './updateT';
export * from './correctRGB';
export * from './correctHSV';
