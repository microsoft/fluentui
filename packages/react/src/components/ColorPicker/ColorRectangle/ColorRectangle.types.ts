import * as React from 'react';
import type { ITheme, IStyle } from '../../../Styling';
import type { IRefObject, IStyleFunctionOrObject } from '../../../Utilities';
import type { IColor } from '../../../utilities/color/interfaces';

/**
 * {@docCategory ColorPicker}
 */
export interface IColorRectangle {
  /** Currently selected color. */
  color: IColor;
}

/**
 * {@docCategory ColorPicker}
 */
export interface IColorRectangleProps {
  /**
   * Gets the component ref.
   */
  componentRef?: IRefObject<IColorRectangle>;

  /**
   * Current color of the rectangle.
   */
  color: IColor;

  /**
   * Label of the ColorRectangle for the benefit of screen reader users.
   * @defaultvalue 'Saturation and brightness'
   */
  ariaLabel?: string;

  /**
   * Format string for the color rectangle's current value as read by screen readers.
   * The string must include descriptions and two placeholders for the current values:
   * `{0}` for saturation and `{1}` for value/brightness.
   * @defaultvalue `'Saturation {0} brightness {1}'`
   */
  ariaValueFormat?: string;

  /**
   * Detailed description for how to use the color rectangle. Moving the thumb horizontally adjusts
   * saturation and moving it vertically adjusts value (essentially, brightness).
   * @defaultvalue 'Use left and right arrow keys to set saturation. Use up and down arrow keys to set brightness.'
   */
  ariaDescription?: string;

  /**
   * Minimum width and height.
   */
  minSize?: number;

  /**
   * Additional CSS class(es) to apply to the ColorRectangle.
   */
  className?: string;

  /**
   * Theme (provided through customization).
   */
  theme?: ITheme;

  /**
   * Call to provide customized styling that will layer on top of the variant rules.
   */
  styles?: IStyleFunctionOrObject<IColorRectangleStyleProps, IColorRectangleStyles>;

  /**
   * Callback for when the color changes.
   */
  onChange?: (ev: React.MouseEvent | React.KeyboardEvent, color: IColor) => void;
}

/**
 * {@docCategory ColorPicker}
 */
export interface IColorRectangleStyleProps {
  /**
   * Theme (provided through customization).
   */
  theme: ITheme;

  /**
   * Additional CSS class(es) to apply to the ColorRectangle.
   */
  className?: string;

  /**
   * Minimum width and height.
   */
  minSize?: number;
}

/**
 * {@docCategory ColorPicker}
 */
export interface IColorRectangleStyles {
  /**
   * Style set for the root element.
   */
  root?: IStyle;

  /**
   * Style set for the light-colored rectangle.
   */
  light?: IStyle;

  /**
   * Style set for the dark-colored rectangle.
   */
  dark?: IStyle;

  /**
   * Style set for the draggable thumb element.
   */
  thumb?: IStyle;

  /**
   * Style for a hidden detailed description for screen reader users.
   */
  description?: IStyle;
}
