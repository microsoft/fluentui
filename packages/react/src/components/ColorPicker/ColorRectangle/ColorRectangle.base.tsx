import * as React from 'react';
import { classNamesFunction, on, initializeComponentRef, KeyCodes, getId } from '../../../Utilities';
import {
  IColorRectangleProps,
  IColorRectangleStyleProps,
  IColorRectangleStyles,
  IColorRectangle,
} from './ColorRectangle.types';

// These imports are separated to help with bundling
import { IColor } from '../../../utilities/color/interfaces';
import { MAX_COLOR_SATURATION, MAX_COLOR_VALUE } from '../../../utilities/color/consts';
import { getFullColorString } from '../../../utilities/color/getFullColorString';
import { updateSV } from '../../../utilities/color/updateSV';
import { clamp } from '../../../utilities/color/clamp';

const getClassNames = classNamesFunction<IColorRectangleStyleProps, IColorRectangleStyles>();

export interface IColorRectangleState {
  color: IColor;
}

/**
 * {@docCategory ColorPicker}
 */
export class ColorRectangleBase
  extends React.Component<IColorRectangleProps, IColorRectangleState>
  implements IColorRectangle {
  public static defaultProps: Partial<IColorRectangleProps> = {
    minSize: 220,
    ariaLabel: 'Saturation and brightness',
    ariaValueFormat: 'Saturation {0} brightness {1}',
    ariaDescription: 'Use left and right arrow keys to set saturation. Use up and down arrow keys to set brightness.',
  };

  private _disposables: (() => void)[] = [];
  private _root = React.createRef<HTMLDivElement>();
  private _isAdjustingSaturation: boolean = true;
  private _descriptionId = getId('ColorRectangle-description');

  constructor(props: IColorRectangleProps) {
    super(props);

    initializeComponentRef(this);

    this.state = { color: props.color };
  }

  public get color(): IColor {
    return this.state.color;
  }

  public componentDidUpdate(
    prevProps: Readonly<IColorRectangleProps>,
    prevState: Readonly<IColorRectangleState>,
  ): void {
    // if props changed (as opposed to a state update), set the value
    // TODO: switch to strict controlled pattern instead
    if (prevProps !== this.props && this.props.color) {
      this.setState({ color: this.props.color });
    }
  }

  public componentWillUnmount() {
    this._disposeListeners();
  }

  public render(): JSX.Element {
    const { minSize, theme, className, styles, ariaValueFormat, ariaLabel, ariaDescription } = this.props;
    const { color } = this.state;

    const classNames = getClassNames(styles!, {
      theme: theme!,
      className,
      minSize,
    });

    const valueText = ariaValueFormat!.replace('{0}', String(color.s)).replace('{1}', String(color.v));

    return (
      <div
        ref={this._root}
        tabIndex={0}
        className={classNames.root}
        style={{ backgroundColor: getFullColorString(color) }}
        onMouseDown={this._onMouseDown}
        onKeyDown={this._onKeyDown}
        role="slider"
        // Proper description of the current hue *and* saturation
        // (screen reader typically chooses this over aria-valuenow)
        aria-valuetext={valueText}
        // Narrator reads aria-valuetext first, but it also reads aria-valuenow after a pause, and
        // per the aria spec this defaults to halfway between min/max (50) if we don't provide it.
        // So provide the value of the most recently adjusted thing.
        aria-valuenow={this._isAdjustingSaturation ? color.s : color.v}
        aria-valuemin={0}
        aria-valuemax={MAX_COLOR_VALUE}
        aria-label={ariaLabel}
        aria-describedby={this._descriptionId}
        data-is-focusable={true}
      >
        <div className={classNames.description} id={this._descriptionId}>
          {ariaDescription}
        </div>
        <div className={classNames.light} />
        <div className={classNames.dark} />
        <div
          className={classNames.thumb}
          style={{ left: color!.s + '%', top: MAX_COLOR_VALUE - color!.v + '%', backgroundColor: color!.str }}
        />
      </div>
    );
  }

  private _onKeyDown = (ev: React.KeyboardEvent<HTMLElement>): void => {
    const { color } = this.state;
    let { s, v } = color;

    const increment = ev.shiftKey ? 10 : 1;

    // Intentionally DO NOT flip the color picker in RTL: its orientation is not very meaningful,
    // and getting all the math and styles flipped correctly is tricky
    // eslint-disable-next-line deprecation/deprecation
    switch (ev.which) {
      case KeyCodes.up: {
        this._isAdjustingSaturation = false;
        v += increment; // V = 100 (lightest) is at the top
        break;
      }
      case KeyCodes.down: {
        this._isAdjustingSaturation = false;
        v -= increment; // V = 0 (darkest) is at the bottom
        break;
      }
      case KeyCodes.left: {
        this._isAdjustingSaturation = true;
        s -= increment;
        break;
      }
      case KeyCodes.right: {
        this._isAdjustingSaturation = true;
        s += increment;
        break;
      }
      default:
        return;
    }

    this._updateColor(ev, updateSV(color, clamp(s, MAX_COLOR_SATURATION), clamp(v, MAX_COLOR_VALUE)));
  };

  private _updateColor(ev: MouseEvent | KeyboardEvent | React.MouseEvent | React.KeyboardEvent, color: IColor): void {
    const { onChange } = this.props;

    const oldColor = this.state.color;
    if (color.s === oldColor.s && color.v === oldColor.v) {
      return; // no change
    }

    if (onChange) {
      onChange(ev as React.MouseEvent | React.KeyboardEvent, color);
    }

    if (!ev.defaultPrevented) {
      this.setState({ color });
      ev.preventDefault();
    }
  }

  private _onMouseDown = (ev: React.MouseEvent): void => {
    this._disposables.push(
      on(window, 'mousemove', this._onMouseMove as (ev: MouseEvent) => void, true),
      on(window, 'mouseup', this._disposeListeners, true),
    );

    this._onMouseMove(ev);
  };

  private _onMouseMove = (ev: MouseEvent | React.MouseEvent): void => {
    if (!this._root.current) {
      return;
    }

    // Leaving the following commented code which is sometimes necessary for debugging:
    // If the primary button (1) isn't pressed, the user is no longer dragging, so turn off
    // the event handlers and exit.
    // if (!(ev.buttons & 1)) {
    //   this._disposeListeners();
    //   return;
    // }

    const newColor = _getNewColor(ev, this.state.color, this._root.current);
    if (newColor) {
      this._updateColor(ev, newColor);
    }
  };

  private _disposeListeners = (): void => {
    this._disposables.forEach(dispose => dispose());
    this._disposables = [];
  };
}

/**
 * Exported for testing only.
 * @internal
 */
export function _getNewColor(
  ev: MouseEvent | React.MouseEvent,
  prevColor: IColor,
  root: HTMLElement,
): IColor | undefined {
  const rectSize = root.getBoundingClientRect();

  const sPercentage = (ev.clientX - rectSize.left) / rectSize.width;
  const vPercentage = (ev.clientY - rectSize.top) / rectSize.height;

  return updateSV(
    prevColor,
    clamp(Math.round(sPercentage * MAX_COLOR_SATURATION), MAX_COLOR_SATURATION),
    clamp(Math.round(MAX_COLOR_VALUE - vPercentage * MAX_COLOR_VALUE), MAX_COLOR_VALUE),
  );
}
