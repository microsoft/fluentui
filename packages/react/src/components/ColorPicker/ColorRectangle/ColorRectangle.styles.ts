import { IColorRectangleStyleProps, IColorRectangleStyles } from './ColorRectangle.types';
import { HighContrastSelector, getHighContrastNoAdjustStyle } from '../../../Styling';
import { IsFocusVisibleClassName } from '../../../Utilities';
import { hiddenContentStyle } from '@fluentui/style-utilities';

export const getStyles = (props: IColorRectangleStyleProps): IColorRectangleStyles => {
  const { className, theme, minSize } = props;
  const { palette, effects } = theme;

  return {
    root: [
      'ms-ColorPicker-colorRect',
      {
        position: 'relative',
        marginBottom: 8,
        border: `1px solid ${palette.neutralLighter}`,
        borderRadius: effects.roundedCorner2,
        minWidth: minSize,
        minHeight: minSize,
        outline: 'none',

        selectors: {
          [HighContrastSelector]: {
            ...getHighContrastNoAdjustStyle(),
          },

          [`.${IsFocusVisibleClassName} &:focus`]: {
            outline: `1px solid ${palette.neutralSecondary}`,
          },
        },
      },
      className,
    ],
    light: [
      'ms-ColorPicker-light',
      {
        position: 'absolute',
        left: 0,
        right: 0,
        top: 0,
        bottom: 0,
        // Intentionally DO NOT flip the color picker in RTL: its orientation is not very meaningful,
        // and getting all the math and styles flipped correctly is tricky
        background: 'linear-gradient(to right, white 0%, transparent 100%) /*@noflip*/',
      },
    ],
    dark: [
      'ms-ColorPicker-dark',
      {
        position: 'absolute',
        left: 0,
        right: 0,
        top: 0,
        bottom: 0,
        background: 'linear-gradient(to bottom, transparent 0, #000 100%)',
      },
    ],
    thumb: [
      'ms-ColorPicker-thumb',
      {
        position: 'absolute',
        width: 20,
        height: 20,
        background: 'white',
        border: `1px solid ${palette.neutralSecondaryAlt}`,
        borderRadius: '50%',
        boxShadow: effects.elevation8,
        transform: 'translate(-50%, -50%)',
        selectors: {
          ':before': {
            position: 'absolute',
            left: 0,
            right: 0,
            top: 0,
            bottom: 0,
            border: `2px solid ${palette.white}`,
            borderRadius: '50%',
            boxSizing: 'border-box',
            content: '""',
          },
        },
      },
    ],
    description: hiddenContentStyle,
  };
};
