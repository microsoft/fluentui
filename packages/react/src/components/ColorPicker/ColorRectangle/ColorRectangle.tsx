import * as React from 'react';
import { styled } from '../../../Utilities';
import { ColorRectangleBase } from './ColorRectangle.base';
import { getStyles } from './ColorRectangle.styles';
import { IColorRectangleProps, IColorRectangleStyles, IColorRectangleStyleProps } from './ColorRectangle.types';

export const ColorRectangle: React.FunctionComponent<IColorRectangleProps> = styled<
  IColorRectangleProps,
  IColorRectangleStyleProps,
  IColorRectangleStyles
>(ColorRectangleBase, getStyles, undefined, { scope: 'ColorRectangle' });
