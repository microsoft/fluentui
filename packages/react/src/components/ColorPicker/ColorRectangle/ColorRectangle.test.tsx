import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { mount, ReactWrapper } from 'enzyme';

import { ColorRectangle } from './ColorRectangle';
import { ColorRectangleBase, IColorRectangleState, _getNewColor } from './ColorRectangle.base';
import { IColorRectangleProps } from './ColorRectangle.types';
import { getColorFromString, IColor } from '../../../utilities/color/colors';
import { KeyCodes } from '../../../Utilities';

describe('ColorRectangle', () => {
  let wrapper: ReactWrapper<IColorRectangleProps, IColorRectangleState, ColorRectangleBase> | undefined;
  let component: renderer.ReactTestRenderer | undefined;
  let colorRectangle: ColorRectangleBase | null = null;
  const colorRectRef = (ref: ColorRectangleBase | null) => {
    colorRectangle = ref;
  };
  const getBoundingClientRect = (size: number, offset: number = 0) => (): DOMRect =>
    ({
      left: offset,
      top: offset,
      right: size + offset,
      bottom: size + offset,
      width: size,
      height: size,
    } as DOMRect);

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
    if (component) {
      component.unmount();
      component = undefined;
    }
  });

  it('renders correctly', () => {
    component = renderer.create(<ColorRectangle color={getColorFromString('#abcdef')!} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('uses provided color', () => {
    wrapper = mount(<ColorRectangle color={getColorFromString('#abcdef')!} componentRef={colorRectRef} />);

    expect(colorRectangle!.color.hex).toEqual('abcdef');
  });

  it('respects color prop change', () => {
    const onChange = jest.fn();
    wrapper = mount(
      <ColorRectangle color={getColorFromString('#abcdef')!} onChange={onChange} componentRef={colorRectRef} />,
    );

    wrapper.setProps({ color: getColorFromString('#AEAEAE')! });
    expect(colorRectangle!.color.hex).toEqual('aeaeae');
    // shouldn't call onChange when the consumer updates the color prop
    expect(onChange).toHaveBeenCalledTimes(0);
  });

  it('uses default aria values', () => {
    const color = getColorFromString('#abcdef')!;
    wrapper = mount(<ColorRectangle color={color} />);

    const element = wrapper.getDOMNode();
    expect(element.getAttribute('aria-label')).toBe('Saturation and brightness');
    expect(element.getAttribute('aria-valuetext')).toBe(`Saturation ${color.s} brightness ${color.v}`);

    const descriptionId = element.getAttribute('aria-describedby');
    expect(descriptionId).toBeTruthy();
    const descriptionEl = element.querySelectorAll('#' + descriptionId)[0];
    expect(descriptionEl).toBeTruthy();
    expect(descriptionEl.textContent).toBe(
      'Use left and right arrow keys to set saturation. Use up and down arrow keys to set brightness.',
    );
  });

  it('uses custom aria values', () => {
    const color = getColorFromString('#abcdef')!;
    wrapper = mount(
      <ColorRectangle
        color={color}
        ariaLabel="custom label"
        ariaValueFormat="v {1} s {0}"
        ariaDescription="custom description"
      />,
    );

    const element = wrapper.getDOMNode();
    expect(element.getAttribute('aria-label')).toBe('custom label');
    expect(element.getAttribute('aria-valuetext')).toBe(`v ${color.v} s ${color.s}`);

    const descriptionId = element.getAttribute('aria-describedby');
    const descriptionEl = element.querySelectorAll('#' + descriptionId)[0];
    expect(descriptionEl.textContent).toBe('custom description');
  });

  it('correctly calculates new color', () => {
    const prevColor = getColorFromString('#fff')!;
    const fakeRoot = { getBoundingClientRect: getBoundingClientRect(100) } as HTMLElement;
    // fake mouse move in top right => red
    const fakeEvent = { clientX: 100, clientY: 0 } as React.MouseEvent<HTMLElement>;

    expect(_getNewColor(fakeEvent, prevColor, fakeRoot)).toEqual(getColorFromString('#ff0000'));
  });

  it('correctly calculates new color when mouse move is out of max bounds', () => {
    const prevColor = getColorFromString('#fff')!;
    const fakeRoot = { getBoundingClientRect: getBoundingClientRect(100) } as HTMLElement;
    // fake mouse move out of bounds to bottom right => black
    const fakeEvent = { clientX: 200, clientY: 200 } as React.MouseEvent<HTMLElement>;
    // by default, black has s=0, but in this case it will be s=100
    const newColor = {
      ...getColorFromString('#000000'),
      s: 100,
    };

    expect(_getNewColor(fakeEvent, prevColor, fakeRoot)).toEqual(newColor);
  });

  it('correctly calculates new color when mouse move is out of min bounds', () => {
    const prevColor = getColorFromString('#000')!;
    const fakeRoot = { getBoundingClientRect: getBoundingClientRect(100, 50) } as HTMLElement;
    // fake mouse move out of bounds to top left => white
    const fakeEvent = { clientX: 0, clientY: 0 } as React.MouseEvent<HTMLElement>;

    expect(_getNewColor(fakeEvent, prevColor, fakeRoot)).toEqual(getColorFromString('#ffffff')!);
  });

  it('preserves hue when changing color to white', () => {
    const prevColor = getColorFromString('#abcdef')!;
    const fakeRoot = { getBoundingClientRect: getBoundingClientRect(100) } as HTMLElement;
    const fakeEvent = { clientX: 0, clientY: 0 } as React.MouseEvent<HTMLElement>;
    const newColor = {
      ...getColorFromString('#ffffff'),
      h: prevColor.h,
    };

    expect(_getNewColor(fakeEvent, prevColor, fakeRoot)).toEqual(newColor);
  });

  it('handles key events', () => {
    const color1 = getColorFromString('#abcdef')!;
    let color: IColor | undefined;
    const onChange = jest.fn((ev: any, newColor: IColor) => {
      color = newColor;
    });
    wrapper = mount(<ColorRectangle color={color1} onChange={onChange} componentRef={colorRectRef} />);
    wrapper.getDOMNode().getBoundingClientRect = getBoundingClientRect(100);

    const checkResult = (s: number, v: number, onChangeCount: number) => {
      expect(onChange).toHaveBeenCalledTimes(onChangeCount);
      expect(color!.s).toBe(s);
      expect(color!.v).toBe(v);
      expect(colorRectangle!.color.s).toBe(s);
      expect(colorRectangle!.color.v).toBe(v);
    };

    wrapper.simulate('keydown', { which: KeyCodes.left });
    checkResult(color1.s - 1, color1.v, 1);

    wrapper.simulate('keydown', { which: KeyCodes.right });
    checkResult(color1.s, color1.v, 2);

    wrapper.simulate('keydown', { which: KeyCodes.up });
    checkResult(color1.s, color1.v + 1, 3);

    wrapper.simulate('keydown', { which: KeyCodes.down });
    checkResult(color1.s, color1.v, 4);

    wrapper.simulate('keydown', { which: KeyCodes.down, shiftKey: true });
    checkResult(color1.s, color1.v - 10, 5);
  });

  it('ignores key events that put value out of range', () => {
    const onChange = jest.fn();
    wrapper = mount(
      <ColorRectangle color={getColorFromString('#fff')!} onChange={onChange} componentRef={colorRectRef} />,
    );
    const initialColor = colorRectangle!.color;

    // white is at top left corner, so going up isn't valid
    wrapper.simulate('keydown', { which: KeyCodes.up });
    expect(colorRectangle!.color).toEqual(initialColor);
    expect(onChange).toHaveBeenCalledTimes(0);
  });

  it('does not modify value if event default prevented', () => {
    let updatedColor: IColor | undefined;
    const onChange = jest.fn((ev: React.SyntheticEvent, newColor: IColor) => {
      updatedColor = newColor;
      ev.preventDefault();
    });
    wrapper = mount(
      <ColorRectangle color={getColorFromString('#fff')!} onChange={onChange} componentRef={colorRectRef} />,
    );
    const initialColor = colorRectangle!.color;

    wrapper.simulate('keydown', { which: KeyCodes.down });
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(updatedColor!.v).toBe(99);
    expect(colorRectangle!.color).toEqual(initialColor);
  });

  it('handles mouse events', () => {
    const onChange = jest.fn();
    wrapper = mount(
      <ColorRectangle color={getColorFromString('#fff')!} onChange={onChange} componentRef={colorRectRef} />,
    );
    wrapper.getDOMNode().getBoundingClientRect = getBoundingClientRect(100);

    // click in top right => red
    const red = getColorFromString('#ff0000')!;
    wrapper.simulate('mousedown', { type: 'mousedown', clientX: 100, clientY: 0, buttons: 1 });
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(colorRectangle!.color).toEqual(red);

    // move farther to right => ignored (keep previous)
    wrapper.simulate('mousemove', { type: 'mousemove', clientX: 200, clientY: 0, buttons: 1 });
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(colorRectangle!.color).toEqual(red);

    // mouse up => keep color
    wrapper.simulate('mouseup');
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(colorRectangle!.color).toEqual(red);
  });
});
