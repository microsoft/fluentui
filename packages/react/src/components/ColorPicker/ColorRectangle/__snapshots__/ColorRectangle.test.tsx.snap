// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorRectangle renders correctly 1`] = `
<div
  aria-describedby="ColorRectangle-description0"
  aria-label="Saturation and brightness"
  aria-valuemax={100}
  aria-valuemin={0}
  aria-valuenow={28}
  aria-valuetext="Saturation 28 brightness 94"
  className=
      ms-ColorPicker-colorRect
      {
        border-radius: 2px;
        border: 1px solid #f3f2f1;
        margin-bottom: 8px;
        min-height: 220px;
        min-width: 220px;
        outline: none;
        position: relative;
      }
      @media screen and (-ms-high-contrast: active), (forced-colors: active){& {
        -ms-high-contrast-adjust: none;
        forced-color-adjust: none;
      }
      .ms-Fabric--isFocusVisible &:focus {
        outline: 1px solid #605e5c;
      }
  data-is-focusable={true}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="slider"
  style={
    Object {
      "backgroundColor": "#0080ff",
    }
  }
  tabIndex={0}
>
  <div
    className=

        {
          border: 0px;
          height: 1px;
          margin-bottom: -1px;
          margin-left: -1px;
          margin-right: -1px;
          margin-top: -1px;
          overflow: hidden;
          padding-bottom: 0px;
          padding-left: 0px;
          padding-right: 0px;
          padding-top: 0px;
          position: absolute;
          white-space: nowrap;
          width: 1px;
        }
    id="ColorRectangle-description0"
  >
    Use left and right arrow keys to set saturation. Use up and down arrow keys to set brightness.
  </div>
  <div
    className=
        ms-ColorPicker-light
        {
          background: linear-gradient(to right, white 0%, transparent 100%) /*@noflip*/;
          bottom: 0px;
          left: 0px;
          position: absolute;
          right: 0px;
          top: 0px;
        }
  />
  <div
    className=
        ms-ColorPicker-dark
        {
          background: linear-gradient(to bottom, transparent 0, #000 100%);
          bottom: 0px;
          left: 0px;
          position: absolute;
          right: 0px;
          top: 0px;
        }
  />
  <div
    className=
        ms-ColorPicker-thumb
        {
          background: white;
          border-radius: 50%;
          border: 1px solid #8a8886;
          box-shadow: 0 3.2px 7.2px 0 rgba(0, 0, 0, 0.132), 0 0.6px 1.8px 0 rgba(0, 0, 0, 0.108);
          height: 20px;
          position: absolute;
          transform: translate(-50%, -50%);
          width: 20px;
        }
        &:before {
          border-radius: 50%;
          border: 2px solid #ffffff;
          bottom: 0px;
          box-sizing: border-box;
          content: "";
          left: 0px;
          position: absolute;
          right: 0px;
          top: 0px;
        }
    style={
      Object {
        "backgroundColor": "#abcdef",
        "left": "28%",
        "top": "6%",
      }
    }
  />
</div>
`;
